---
title: 异步
date: 2020-08-17 21:07:31
permalink: /pages/459f9f/
categories:
  - 大前端
  - js基础
---
# 异步编程

[toc]

众所周知，目前主流的 JavaScript 程序都是以单线程的形式执行的，JavaScript 单线程执行的原因和它的设计初衷有关，最早这门语言就是运行在浏览器上的脚本语言，主要用于页面上的动态交互，而实现页面交互的核心就是 DOM 操作，这就决定了它必须使用单线程模型，否则就会出现很复杂的线程同步问题。试想在 JavaScript 中有多个线程同时工作，其中一个线程修改了某个元素，而另外一个线程又同时删除了这个元素，那浏览器就没法确定以哪个线程为准，为了避免这种线程同步问题，JavaScript 最开始就被设计成了单线程，这也就成了这门语言的核心特点之一。这里说的单线程是指在 JavaScript 执行环境中，负责执行代码的线程只有一个，如果有多个任务就必须排队，一个个依次执行，这种模式的优点就是更安全、更简单，缺点就是遇到遇到耗时任务后面的任务都需要排队等待，可能导致程序出现假死的情况。

为了解决耗时任务的执行，JavaScript 将任务的执行模式分成了两种，分别是同步模式和异步模式和异步模式。

## 同步模式和异步模式的差异

**同步模式：**

代码中的任务依次执行，后一个必须等待前一个任务结束才能执行，程序的执行顺序和代码的编写顺序是一致的，这种方式会比较简单，JavaScript 在单线程模式下大多数任务都会按照同步模式执行。

```javascript
  console.log('global begin')
  function fun1 () {
    console.log('fun1')
  }
  function fun2 () {
    console.log('fun2')
    fun1()
  }
  fun2()
  console.log('global end')
  // 结果：
  //  global begin
  //  fun2
  //  fun1
  //  global end
```

执行时 JS 引擎会把整个代码加载进来，在调用栈中压入一个匿名的调用（把全部代码放入匿名函数中执行），然后逐行执行代码：

- 第一行遇到 ``console.log('global begin')``，把这一行压入调用栈执行，控制台打印 ``global begin``，这一行弹出调用栈，然后继续执行；

- 接着是两个函数的声明，不管函数还是变量的声明都不会产生调用，程序继续往下执行；

- 再往下就是 ``fun2()`` 调用，引擎把 fun2 压入调用栈，然后逐行执行 fun2：
  - 最开始把``console.log('fun2')`` 压入调用栈，打印了一条消息，接着又把 ``console.log('fun2')`` 弹出；
  - 把 fun1 函数压栈，接着执行完毕，弹出 fun1;
  - 接着弹出 fun2 ；

- 最后打印 ``global end``，等到调用结束，调用栈清空；

如果某一行执行时间过长，就会出现阻塞，对用户而言就会出现界面卡顿，耗时操作需要通过异步来解决（比如 ajax/文件读写）。

**异步模式：**

异步模式的 api 不会等待当前任务结束才开始执行下一个任务，对于耗时操作都是开启之后就立即往后执行下一个任务，耗时任务的后续逻辑一般通过回调函数的方式定义，等异步任务执行结束后就会调用回调函数。异步模式对 JavaScript 非常重要，如果没有异步模式单线程的 JavaScript 语言就无法同时处理大量耗时任务，对于开发人员来说，异步模式下的代码执行顺序会出现跳跃，不会像同步代码那样通俗易懂。

```javascript
  console.log('global begin')
  setTimeout(function timeout1(){
    console.log('timeout1')
  },1800)
  setTimeout(function timeout2(){
    console.log('timeout2')
    setTimeout(function inner(){
      console.log('inner timeout')
    },1000)
  },1000)
  console.log('global end')
```

- 最开始也是加载整体代码，然后压入匿名全局调用；
- 压栈、执行、弹出 ``console.log('global begin')``
- 压入 ``setTimeout(function(){ console.log('timer1') },1000)``，以 Web 浏览器为例，Web API 为 setTimeOut 开启一个倒计时器（单独工作，不受 js 线程影响），弹出 setTimeOut ，继续往下；
- 再压入第二个 setTimeOut，同样压栈、开启另一个倒计时器，然后弹栈；
- 最后压栈、执行、弹出 ``console.log('global end')``，整体匿名调用完成，清空调用栈；
- 调用栈清空了，eventLoop 发挥作用，负责监听调用栈和事件队列（一旦调用栈结束了，就从事件队列中取出第一个回调函数，然后引入调用栈）；

  - timer2 计时器先结束，然后 timeout2 函数被放入到事件队列的第一位；
  - 然后 timer1 结束，timeout1 被放入事件队列的第二位；
  - 一旦事件队列发生变化，事件循环就会把事件队列的第一个函数（timeout2）取出来，压入调用栈执行，开启新的一轮执行，执行过程和之前一致，如果遇到异步调用也是相同的情况；
  - 往后就是不断地调用重复，直到调用栈和事件队列都没有需要执行的任务了，那整体代码结束；

![异步调用](https://gitee.com/leixiaoai/markdown/raw/master/01.大前端/02.js基础/images/异步/异步调用.png)

如图，JavaScript 主线程某个时刻发起异步调用，然后继续往后执行其它任务，同时异步线程会去执行异步任务，异步任务执行完成之后会把异步任务放入事件队列，js 主线程完成所有任务之后，会依次执行事件队列的任务。需要注意的是 JavaScript 是单线程的，而浏览器不是单线程的，具体来说就是 JavaScript 通过调用了浏览器的 API 来实现异步调用。

## 回调函数

异步模式对单线程的 JavaScript 非常重要，同时也是 JavaScript 的核心特点，也是因为异步模式 API 的关系，写出来的代码不容易读，执行顺序也会复杂很多，自从 es2015 开始新增的 es 语法慢慢弥补了 JavaScript 的不足。

回调函数：

这是 JavaScript 异步编程的根本方式，其实所有异步函数的根本都是回调函数，回调函数可以理解为知道要执行什么事情，但是不知道这个事情依赖的任务什么时候完成，所以说最好的办法就是把这些事情的步骤写到一个函数（回调函数）当中，交给任务的执行者，执行者是知道什么时候结束，等结束之后由异步任务的执行者执行。以 ajax 为例，ajax 就是希望拿到数据之后去做一些处理，但是请求什么时候完成不知道，所以得把请求响应之后要执行的任务放到函数中，ajax 执行完成之后会自动执行这些任务，这种由调用者定义，交给执行者执行的函数就被称之为回调函数。具体方法也很简单，就是把函数作为参数传递，只不过这种方式不利于阅读，而且执行顺序也会非常混乱。

## Promise

回调函数是 JavaScript 异步编程的根基，但是直接使用传统回调函数的方式，去完成复杂的异步流程，就无法避免大量的回调函数嵌套，这就会导致常说的回调地狱问题，为了避免回调地狱的问题 CommonJS 社区就提出了 Promise 规范，目的就是为 JavaScript 提供一种更合理、更强大的异步编程方案，后来在 es2015 中被标准化，称为语言规范。

所谓的 Promise 就是用一个对象表示异步任务结束之后是成功还是失败，就像是内部对外部做出了一个承诺，最开始承诺是待定状态（padding），最终可能成功（Fulfilled），也有可能失败（rejected），不管是达成还是失败，都会有相应的反应 onFulFilled、onRejected，在承诺结束之后都会有相应的任务被执行，而且还有一个明显的特点，就是一旦明确了结果就不可以被改变了。

```javascript

  // Promise 的构造函数需要接受一个函数作为参数，这个函数就可以理解为对象承诺的逻辑
  // 这个函数会在构造 Promise 的过程中被同步执行
  // 这个函数接受两个参数 resolve/reject，分别将 Promise 的状态改为成功、失败
  // 状态是确定的，也就是说只会调用 resolve/reject 之一
  const promise = new Promise(function(resolve,reject){
    console.log('start')
    resolve(100)
    reject(new Error('rejected')) // 一般传递一个错误对象，表示为什么失败
  })

  promise.then(function(value){
    console.log('resolve',value)
  },function(error){
    console.log('error',error)
  })
  console.log('end')
  // start
  // end
  // resolve 100
```

Promsie 封装 ajax（需要在浏览器执行） ：

```javascript
  function ajax (url) {
    return new Promise(function (resolve, reject) {
      const xhr = new XMLHttpRequest()
      xhr.open('GET', url)
      xhr.responseType = 'json' // html5 新特性，获取 json 对象
      xhr.onload = function () { //  html5 新特性，readystate === 4 请求完成之后
        if (xhr.status === 200) {
          resolve(this.response)
        } else {
          reject(new Error(this.statusText))
        }
      }
      xhr.send()
    })
  }
  ajax('./').then(function (res) {
    console.log(res)
  }, function (err) {
    console.log(err)
  })
```

**链式调用：**

从前面来看 Promise 的本质也就是使用回调函数的方式，来定义异步任务结束之后要执行的任务，只不过这里的回调函数是通过 then 方法传递进去的，而且 Promise 将回调分成了成功和失败两种，如果需要串联执行多个异步任务，还是会出现回调嵌套的问题。例如需要先获取 user 信息，再查询 order 信息，这样 Promise 不就没有任何意义了，反而还增加了额外的代码？其实嵌套使用 Promise 是使用 Promise 常见的误区，这个问题可以通过使用 Promsie then 链式调用的方式解决，保证异步任务的扁平化。

相比传统回调的优势，Promise 最大的优势就是可以链式调用，这样就可以避免回调嵌套，Promise 有 then 和 catch 两个方法，then 最大的特点就是会返回一个 Promise。

```javascript
  function ajax (url) {
    return new Promise(function (resolve, reject) {
      const xhr = new XMLHttpRequest()
      xhr.open('GET', url)
      xhr.responseType = 'json' // html5 新特性，获取 json 对象
      xhr.onload = function () { //  html5 新特性，readystate === 4 请求完成之后
        if (xhr.status === 200) {
          resolve(this.response)
        } else {
          reject(new Error(this.statusText))
        }
      }
      xhr.send()
    })
  }
  const promise1 = ajax('./')
  const promise2 = promise1.then(function (res) {
  }, function (err) {
  })
  // 以往的链式调用都会返回 this，但是 Promise 不一样
  // then 方法返回的是一个全新的 Promise 对象，用于实现 Promise 链条
  // 每个 Promise 都可以实现一个新的异步任务，每个 then 方法都是在为上一个 Promise 添加状态明确的回调
  console.log(promise2 === promise1) // false
```

- Promise 的 then 方法会返回一个全新的 Promise 对象，可以通过链式调用的方式添加 then 方法；
- 链式调用后面的 then 方法，就是为上一个 then 返回的 Promise 注册回调；
- 前面 then 方法的返回值会作为后面 then 方法的参数；
- 如果返回 Promise，后面的 then 就会等待这个 Promise 结束；

**异常处理：**

正如前面所说 Promise 一旦调用失败、then 方法执行异常，就会触发 onRejected 回调函数，onRejected 在处理失败或者请求失败都会被执行，关于 onRejected 的注册还有一个常见的办法就是使用 catch 方法，这种方法更加适用于链式调用。

```javascript
  const promise1 = ajax('./').then(function (res) {
    return ajax('/ada/dad')
  }, function (err) {
    // 捕捉不到 /ada/dad 的异常
    console.log(err)
  }).catch(function(err){
    // 可以捕捉到 /ada/dad 的异常
    console.log('catch',err) // catch Error: Not Found at XMLHttpRequest.xhr.onload
  })
```

另外还可以全局注册 unhandledrejection 事件来捕获没有被代码捕获的异常，浏览器注册在 window 上，nodejs 注册到 process 中（这种方式不推荐使用，应该在代码层面明确捕获每个异常）：

```javascript
  process.on('unhandledRejection',(reason,promise)=>{
    console.log(reason,promise)
  })
```

**Promise 静态方法：**

- Promise.resolve()：

  - 把一个值转换为 Promise 对象；

    ```javascript
      Promise.resolve(100).then(res => console.log(res))
    ```
  
  - 如果接受的是 Promise 会原样返回；

    ```javascript
      const timeOutPromise = new Promise(function(resolve,reject){
        setTimeout(()=>{
          console.log('time out')
        },500)
      })
      const promise = Promise.resolve(timeOutPromise)
      console.log(promise === timeOutPromise) // true
    ```

  - 如果传入有 then 方法的对象，这样的对象（实现了 thenable 接口）也可以作为 Promise 执行。原生 Promise 没有普及之前，很多都使用第三方库实现，第三方 Promise 转换为原生的 Promise 就可以借助 Promise.resolve 机制；

    ```javascript
      Promise.resolve({
        then (onFulFilled, onRejected) {
          onFulFilled('then')
        }
      }).then(value => {
        console.log('then1',value)
      }) // then1 then
    ```

- Promise.reject 无论传入什么参数，都会作为失败的理由；

**并行执行：**

前面介绍的都是通过 Promise 串联执行多个任务，一个任务结束之后才能开启下一个任务，相比于传统的异步编程方式，Promise 提供了更加扁平的异步编程体验，如果需要并行执行异步任务，Promise 也提供了更为完善的机制。例如需要请求多个接口，接口没有依赖，最好的方式就是同时请求它们，避免依次请求消耗大量时间，并行请求很容易实现，就是单独调用，但是我们如何判断所有请求都结束了呢？传统编程是定义一个计数器，每结束一个计数器就累加一下，直到计数器的值和任务相等时，就表示所有任务都结束了，这种方法很麻烦，还有考虑出现异常的情况（注意：因为任务队列的存在，不会出现并发问题），如果使用 Promise.all 方法会方便很多。

Promise.all 需要接收的是一个数组，数组每个元素都是一个 Promise 对象，我们可以把这些 Promise 看做一个个的异步任务，这个方法会返回一个全新的 Promise 对象，当内部所有的 Promise 执行完成之后，这个全新的 Promise 才会完成，此时拿到的就是一个数组，在这个数组中包含了每一个任务执行的结果，执行过程中如果有一个 Promise 发生异常，这个全新的 Promise 就会以失败告终，这是一种很好地同步执行 Promise 的方式。

```javascript
  Promise.all([
    new Promise((resolve) => {
      setTimeout(() => {
        resolve('timer1 500')
      }, 500)
    }),
    new Promise((resolve) => {
      setTimeout(() => {
        resolve('timer2 1000')
      }, 1000)
    })
  ]).then(res => {
    console.log(res) // ["timer1 500", "timer2 1000"]
  })
```

另外还有 Promise.race 方法也可以把所有 Promise 组合成全新的 Promise，但是和 Promise.all 不同的是，Promise.race 只会等待第一个任务结束，只要有一个任务完成，新的 Promise 也会完成。可以实现异步任务超时控制：

```javascript
  Promise.race([
    new Promise((resolve) => {
      setTimeout(() => {
        resolve('50 毫秒内没有执行完成')
      }, 50)
    }),
    new Promise((resolve) => {
      setTimeout(() => {
        resolve('timer2 1000')
      }, 1000)
    })
  ]).then(res => {
    console.log(res) // 50 毫秒内没有执行完成
  })
```

**Promise 执行时序：**

即使 Promsie 没有异步代码，Promise 中的代码也得等到同步代码执行完成之后才能执行：

```javascript
  console.log('start')

  setTimeout(() => console.log('timeout'), 0)

  Promise.resolve().then(() => console.log('then1'))
    .then(() => console.log('then2'))
    .then(() => console.log('then3'))

  new Promise((resolve, reject) => {
    console.log('promise2');
    resolve(1)
  }).then(() => console.log('promise2 then1'))
    .then(() => console.log('promise2 then2'))

  console.log('end')
  // 输出 start promise2 end
  // then1
  // promise2 then1
  // then2
  // promise2 then2
  // then3
  // timeout
```

js 有微任务和宏任务之分：

- 微任务在同步代码执行完成之后开始执行（Promise、MutationObserver、node 的 process.$nextTick）；
- 宏任务在微任务结束之后才能执行；

## Generator

**生成器函数：**

相比传统的异步调用，Promise 的最大优势就是可以使用链式调用解决回调嵌套的问题，Promise 的解决办法就是一个 then 处理一个异步任务，最后形成一个扁平的任务链条，实现任务串联执行。但是编写还是会出现大量的回调函数，虽然他们没有嵌套，但是还是没有达到传统代码的可读性，如果采用类似同步的方式编写会更加简洁、容易理解。

生成器函数就是在普通的函数前面多了一个 * 号

- 当我们调用生成器函数时，不会立即执行，而是得到一个生成器对象；
- 直到我们调用生成器对象的 next 方法，这个函数才会开始执行；
- 其次就是可以在函数内部使用 yield 向外返回一个值，另外在返回值中还有一个 done 属性，表示函数是否全部执行完成；
- yield 不会像传统的 return 那样结束函数的执行，只是暂停函数的执行，直到外界下一次调用 next 方法时继续从 yield 的位置执行，
- 如果在 next 函数中传入一个参数，这个参数可以在 yield 左侧接收；
- 如果调用  generator.throw，生成器函数会抛出一个异常，如果不捕获就会直接停止当前执行；

```javascript
  function * test(){
    console.log('test')
    const req1 = yield 1
    console.log(req1)
    try {
      const req2 = yield 2
      console.log(req2)
    } catch (error) {
      console.log(error)
    }
    yield 3
  }
  // 以下不会立即执行
  const generator  = test()
  console.log(generator.next())
  // test
  // {value: 1, done: false}
  console.log(generator.next('传入A'))
  // 传入A
  // {value: 2, done: false}
  generator.throw(new Error('Generator Error'))
  // Error: Generator Error
  //     at eval (eval at <anonymous> (VM1178:8), <anonymous>:17:17)
  //     at VM1179:18
  // {value: 3, done: false}
```

**使用 Generator 管理异步流程：**

可以借助 yield 能够暂停生成器函数的特点，实现更优的异步程序，具体实现方式：

```javascript
  function timeOut (id) {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve({ name: "小明" + id, age: id, id: id })
      }, 1000)
    })
  }
  function* main () {
    // 第一个异步函数以及处理
    const res1 = yield timeOut(1)
    console.log(res1)
    // 第二个异步函数以及处理
    const res2 = yield timeOut(12)
    console.log(res2)
    // 第三个异步函数以及处理
    const res3 = yield timeOut(18)
    console.log(res3)
    // 第四个异步函数以及处理
    const res4 = yield timeOut(24)
    console.log(res4)
  }
  const generator = main()
  const res1 = generator.next()
  res1.value.then(data => {
    const res2 = generator.next(data)
    res2.value.then(data => {
      const res3 = generator.next(data)
      res3.value.then(data => {
        generator.next(data)
      })
    })
  })
  // 依次输出：
  // {name: "小明1", age: 1, id: 1}
  // {name: "小明12", age: 12, id: 12}
  // {name: "小明18", age: 18, id: 18}
```

在上述例子中，我们使用生成器函数把 Promise 的回调函数写在了 yield 后，实现了异步问题的代码简化，对于生成器函数的回调可以使用递归解决：

```javascript
  const g1 = main()
  function handleResult (result) {
    if (result.done) {
      return
    }
    result.value.then(data => {
      handleResult(g1.next(data))
    }, error=>{
      g
    })
  }
  handleResult(g1.next())
  // 依次输出：
  // {name: "小明1", age: 1, id: 1}
  // {name: "小明12", age: 12, id: 12}
  // {name: "小明18", age: 18, id: 18}
  // {name: "小明24", age: 24, id: 24}
```

加上异常处理后，我们可以封装一个 exec 函数用于生成器函数的调用：

```javascript
  function timeOut (id) {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        if (id < 24) {
          resolve({ name: "小明" + id, age: id, id: id })
        } else {
          reject(new Error('ID 超过24'))
        }
      }, 1000)
    })
  }
  function* main () {
    try {
      const res1 = yield timeOut(1)
      console.log(res1)
      const res2 = yield timeOut(12)
      console.log(res2)
      const res3 = yield timeOut(18)
      console.log(res3)
      const res4 = yield timeOut(24)
      console.log(res4)
    } catch (error) {
      console.log(error)
    }
  }
  function exec (fun) {
    const generator = fun()
    function handleResult (result) {
      if (result.done) {
        return
      }
      result.value.then(data => {
        handleResult(generator.next(data))
      }, error => {
        generator.throw(error)
      })
    }
    handleResult(generator.next())
  }
  exec(main)
  // {name: "小明1", age: 1, id: 1}
  // {name: "小明12", age: 12, id: 12}
  // {name: "小明18", age: 18, id: 18}
  // Error: ID 超过24
  //     at <anonymous>:7:16
```

这样的生成器函数执行器在社区中有一个完善的库：<https://github.com/tj/co>

这种方案在之前很流行的，但是自从 es 有了 async await 之后就没那么普及了，不过使用 generator 方案的最大优势就是让我们的异步调用回到扁平化，这是 js 异步编程中很重要的一步。

## Async Await

有了 Generator 之后，JavaScript 中的异步编程与同步代码就有类似的体验了，但是使用 Generator 这种方案还需要手动编写一个执行器函数，比较麻烦，在 es2017 中新增了一个 async await 同样提供了这种扁平化的异步编程体验，而且它是语言层面标准的语法，使用起来更加方便，其实 async await 就是生成器函数的一种更方便的语法糖，我们只要把生成器函数改成普通函数，使用 async 修饰，yield 修改成 await 就可以了：

```javascript
  async function main () {
    try {
      const res1 = await timeOut(1)
      console.log(res1)
      const res2 = await timeOut(12)
      console.log(res2)
      const res3 = await timeOut(18)
      console.log(res3)
      const res4 = await timeOut(24)
      console.log(res4)
    } catch (error) {
      console.log(error)
    }
  }
  main()
```

async 函数的效果也和 Generator 一致，最大的好处就是不需要去配合类似于 co 这样的执行器，因为它是语言层面的标准异步语法，其次就是 async 会返回一个 Promise 对象，方便我们更好地控制代码，还有一个需要注意的是 await 只能出现在函数内部，不能再顶层作用域使用（现在已经在开发了，不就以后可能出现在标准中）。

